base/array.jl

```diff
@@ -1288,6 +1288,41 @@ function insert!(a::Array{T,1}, i::Integer, item) where T
     return a
 end
 
+"""
+    keepat!(a::Vector, inds)
+
+Remove the items at all the indices which are not given by `inds`, and return the modified `a`.
+Items which are kept are shifted to fill the resulting gaps.
+
+`inds` can be either an iterator or a collection of sorted and unique integer indices.
+See also [`deleteat!`](@ref).
+
+# Examples
+```jldoctest
+julia> keepat!([6, 5, 4, 3, 2, 1], 1:2:5)
+3-element Array{Int64,1}:
+ 6
+ 4
+ 2
+```
+"""
+function keepat!(a::Vector, inds)
+    n = length(a)
+    l = 0
+    i = 1
+    for k in inds
+        1 <= k <= n || throw(BoundsError(a, k))
+        l < k || throw(ArgumentError("indices must be unique and sorted"))
+        if i != k
+            @inbounds a[i] = a[k]
+        end
+        l = k
+        i += 1
+    end
+    _deleteend!(a, n-i+1)
+    a
+end
```

Comment:
Should we make this generic and move it to abstractarrays.jl?

```suggestion
function keepat!(a::AbstractVector, inds)
    isempty(inds) && return a
    local prev
    i = firstindex(a)
    for k in inds
        @isdefined(prev) && (prev <= k || throw(ArgumentError("indices must be sorted")))
        if i != k
            a[i] = a[k]
        end
        prev = k
        i = nextind(a, i)
    end
    deleteat!(a, i:lastindex(a))
    return a
end
```

Reply:
I added an `@inbouds` in the generic version to mitigate a bit the slow-down, removed the `isempty` check which was incorrect. I also re-added the constraint to have unique indices, otherwise we have to add another check, unless we would want `keepat!([1, 2, 3], [1, 2, 2]) == [1, 2, 2])` ? For now I prefer the non-ambiguous behavior, which might also catch bugs from the user.